### PR TITLE
PYTHON-3074 Add documentation for type hints

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: doc/examples/type_hints.rst [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  >>> client: MongoClient[Dict[str, Any]] = MongoClient()
+  >>> collection = client.test.test
+  >>> inserted = collection.insert_one({"x": 1, "tags": ["dog", "cat"]})
+  >>> retrieved = collection.find_one({"x": 1})
+  >>> assert isinstance(retrieved, dict)
+
+Client Document Type
+--------------------
+
+:class:`~pymongo.mongo_client.MongoClient` is generic on the document type used to decode BSON documents.
```
Comment (ShaneHarvey) [inline doc/examples/type_hints.rst:59]: We should mention that Client/Database/Collection are all generic too. For instance the TypedDict Movie example would probably make more sense at the collection level:
```python
  >>> from pymongo.collection import Collection
  ...
  >>> client: MongoClient[Dict[str, Any]] = MongoClient()
  >>> collection: Collection[Movie] = client.test.movies
```

File: doc/examples/type_hints.rst [hunk 2]
```diff
@@ -0,0 +1,149 @@
+
+.. _type_hints-example:
+
+Type Hints
```
Comment (ShaneHarvey) [inline doc/examples/type_hints.rst:4]: Are there any common mypy/typing errors that users will likely run into? I suggest we add a section that shows examples of each error and how to resolve it. For example "a type for :class:`~pymongo.mongo_client.MongoClient` must be specified" error would be a good one to start with.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: doc/examples/type_hints.rst (hunk 1 above)
Comment (blink1073) [reply doc/examples/type_hints.rst:59]: Done

File: doc/examples/type_hints.rst (hunk 2 above)
Comment (blink1073) [reply doc/examples/type_hints.rst:4]: Done

File: pymongo/common.py [hunk 3]
```diff
@@ -448,7 +448,15 @@ def validate_document_class(
...
-    if not issubclass(value, (abc.MutableMapping, RawBSONDocument)):
+    # issubclass can raise TypeError for generic aliases like SON[str, Any].
+    # In that case we can use the base class for the comparison.
+    is_mapping = False
+    try:
+        is_mapping = issubclass(value, abc.MutableMapping)
+    except TypeError:
+        if hasattr(value, "__origin__"):
+            is_mapping = issubclass(value.__origin__, abc.MutableMapping)
+    if not is_mapping and not issubclass(value, RawBSONDocument):
```
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/common.py:459]: Done

File: test/test_mypy.py [hunk 4]
```diff
@@ -309,6 +309,9 @@ def test_son_document_type(self) -> None:
         assert retreived is not None
         retreived["a"] = 1
 
+    def test_son_document_type_runtime(self) -> None:
+        client = MongoClient(document_class=SON[str, Any])
```
Comment (ShaneHarvey) [inline test/test_mypy.py:313]: Add connect=False to avoid starting threads.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
one minor comment.

File: test/test_mypy.py (hunk 4 above)
Comment (blink1073) [reply test/test_mypy.py]: Fixed

**Review by ShaneHarvey — APPROVED** [review]

File: doc/examples/type_hints.rst [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+===========
+
+As of version 4.1, PyMongo ships with `type hints`_.
+
+With type hints, Python type checkers can easily find bugs before they reveal themselves in your code.  If your IDE is configured to use type hints,
```
Comment (juliusgeo) [inline doc/examples/type_hints.rst:9]: Maybe put a link here for how to configure type hints in popular IDEs?

Comment (blink1073) [reply doc/examples/type_hints.rst]: Done

**Review by juliusgeo — APPROVED** [review]
LGTM!

**Pull request merged**